custom-gas-token: add e2e tests for activating custom gas token on exising chain and bridging tokens to l2

geoknee opened the pull request:
Needs some work to tidy up, but raising this early to get some early feedback.

Towards https://github.com/ethereum-optimism/client-pod/issues/766

geoknee committed:
add tests for activating custom gas token and bridging tokens to l2

geoknee commented:
Question: I think I may need to edit the circle config to trigger the new test to run in CI?

geoknee commented:
> Question: I think I may need to edit the circle config to trigger the new test to run in CI?

Edit: no, this is not necessary. [Test is running and passing on CI.](https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/52411/workflows/be0e43d7-5701-4f67-bb40-bc129966c728/jobs/2274524/parallel-runs/3?filterBy=ALL)

geoknee commented:
We do have a test failure on the `op-e2e-HTTP-tests-fault-proof` job. From what I can tell, the custom gas token feature is not currently compatible with FPAC. Only the `OptimismPortal` has been updated to enable custom gas tokens, whereas on an FPAC chain we use `OptimismPortal2`.

tynes commented on op-e2e/custom_gas_token_bridge_test.go:
```diff
@@ -0,0 +1,262 @@
+
+}
+
+func waitForTx(t *testing.T, tx *types.Transaction, err error, client *ethclient.Client) {
```
fwiw we do have tooling for this elsewhere in op-e2e, these waits sometimes add flakes

geoknee replied:
This is a thin wrapper around a util from the op-e2e/wait package, so this is at least not reinventing the wheel.

tynes commented:
This all looks sane to me

geoknee committed:
update test following concurrent change in SystemConfig ABI

geoknee committed:
update systemconfig bindings

steps to reproduce:

git checkout  v1.7.3 -- op-bindings

remove WETH9 from artifacts.json

cd op-bindings

make bindgen-generate-local

select this one file discard the others

geoknee committed:
use require.NoError instead of require.Nil

geoknee committed:
add assertions for WETH name and symbol getters

tynes commented:
The `op-upgrade` tool is breaking due to ABI changes. It shouldn't be your responsibility to handle breaking things from other codebases. The fix for this is simple, we can just delete some code

```
golangci-lint run -E goimports,sqlclosecheck,bodyclose,asciicheck,misspell,errorlint --timeout 5m -e "errors.As" -e "errors.Is" ./...
op-chain-ops/upgrades/check.go:1: : # github.com/ethereum-optimism/optimism/op-chain-ops/upgrades
op-chain-ops/upgrades/l1.go:760:4: unknown field L2OutputOracle in struct literal of type "github.com/ethereum-optimism/optimism/op-bindings/bindings".SystemConfigAddresses (typecheck)
package upgrades
op-chain-ops/cmd/op-upgrade/main.go:20:2: could not import github.com/ethereum-optimism/optimism/op-chain-ops/upgrades (-: # github.com/ethereum-optimism/optimism/op-chain-ops/upgrades
op-chain-ops/upgrades/l1.go:760:4: unknown field L2OutputOracle in struct literal of type "github.com/ethereum-optimism/optimism/op-bindings/bindings".SystemConfigAddresses) (typecheck)
	"github.com/ethereum-optimism/optimism/op-chain-ops/upgrades"
	^
op-chain-ops/cmd/op-version-check/main.go:15:2: could not import github.com/ethereum-optimism/optimism/op-chain-ops/upgrades (-: # github.com/ethereum-optimism/optimism/op-chain-ops/upgrades
op-chain-ops/upgrades/l1.go:760:4: unknown field L2OutputOracle in struct literal of type "github.com/ethereum-optimism/optimism/op-bindings/bindings".SystemConfigAddresses) (typecheck)
	"github.com/ethereum-optimism/optimism/op-chain-ops/upgrades"
```

tynes committed:
build: fix

tynes commented:
@geoknee I pushed a commit that fixes the build issue in `op-chain-ops`

tynes commented on op-e2e/custom_gas_token_bridge_test.go:
```diff
@@ -0,0 +1,262 @@
+	addresses.OptimismMintableERC20Factory, err = systemConfig.OptimismMintableERC20Factory(&bind.CallOpts{})
+	require.NoError(t, err)
+
+	// minGasLimit, err := systemConfig.MinimumGasLimit(&bind.CallOpts{})
```
any reason this is commented out still?

tynes commented on op-e2e/custom_gas_token_bridge_test.go:
```diff
@@ -0,0 +1,262 @@
+	waitForTx(t, tx, err, l1Client)
+
+	// Set up a signer which controls the Proxy Admin Owner SAFE
+	cliqueSignerOpts, err := bind.NewKeyedTransactorWithChainID(cfg.Secrets.CliqueSigner, cfg.L1ChainIDBig())
```
I think its by happenstance that the proxy admin owner is the clique signer, we are moving away from using clique for L1 completely. A future PR could rename clique signer, alice, bob etc to be more related to the derivation path of the key

tynes commented on op-e2e/custom_gas_token_bridge_test.go:
```diff
@@ -0,0 +1,262 @@
+	require.NoError(t, err)
+
+	// Encode calldata for upgrading SystemConfigProxy to the StorageSetter implementation
+	proxyAdminABI, err := abi.JSON(strings.NewReader(bindings.ProxyAdminABI))
```
I think there is a way to get the abi with `bindings.ProxyAdminMetadata.GetAbi()`

tynes commented on op-e2e/custom_gas_token_bridge_test.go:
```diff
@@ -0,0 +1,262 @@
+	require.NoError(t, err)
+
+	setCustomGasToken(t, cfg, sys, wethAddress)
+
```
nit: extra newline

tynes commented on op-e2e/custom_gas_token_bridge_test.go:
```diff
@@ -0,0 +1,262 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestCustomGasTokenLockAndMint(t *testing.T) {
```
Do we really need to break this up into 2 tests? feels like there is very little utility in having 2 diff tests. I could see subtests being useful, ie calling up to `setCustomGasToken` in the top level scope then starting subcontexts for specific tests. Writing tests in this manner will help to ensure to reviewers that the test cases are exhaustive

tynes replied:
Basically we want to ensure there are not any reversions in behavior specifically around the weth predeploy since this code will run on OP Mainnet

For non custom gas token chains:

- `OptimismPortal.depositERC20Transaction()` reverts
- `L1Block.gasPayingToken` returns `(ETHER_TOKEN_ADDRESS, 18)`
- `L1Block.gasPayingTokenSymbol` returns `ETH`
- `L1Block.gasPayingTokenName` returns `Ether`
- `SystemConfig.gasPayingToken` returns `(ETHER_TOKEN_ADDRESS, 18)`
- `SystemConfig.gasPayingTokenSymbol` returns `ETH`
- `SystemConfig.gasPayingTokenName` returns `Ether`
- `WETH` predeploy returns `name()` == `"Wrapped Ether"` and `symbol()` returns `"WETH"`

For custom gas token chains:

- basic lock erc20+mint flow
    - validate native asset balance of `to` address changes on L2 by `value`
- basic withdraw flow for erc20s
    - user withdraws native token (call `L2ToL1MessagePasser` directly)
    - gets custom erc20 token on L1
- `WETH` predeploy returns the correct name + symbol, should be prefixed with `Wrapped` and `W`
- `L1Block.gasPayingToken`, `L1Block.gasPayingTokenName`, `L1Block.gasPayingTokenSymbol` all correct, compared against values in ERC20 token

geoknee replied:
Yep these tests are on my radar. I can put them in as subtests in the way you suggest. 

Re: custom gas token chains vs non custom gas token chains. My plan is as follows: we have a test for each, and they run against a chain where the feature is installed but not enabled.
* to test non-CGT chains we can just run our assertions as normal in the first test
* to test CGT chains we simply enable it at runtime at the start of the second test (this is what I have so far here and I will complete it).

Let me know if that sounds wrong to you.

tynes replied:
This all sounds great to me! Thank you

geoknee committed:
refactor using setup + subtests

geoknee committed:
minor changes in response to review

geoknee committed:
devnet-allocs: use sender with known private key

some e2e tests require the ability to transact as the system owner safe owner, so we can't use the foundry default sender (private key unknown)

This uses the same address as "Deployer"

geoknee committed:
use deployer as Safe owner

geoknee committed:
fixmeup

geoknee committed:
WIP refactor test

geoknee committed:
add most of the remaining test assertions

geoknee committed:
rename file

geoknee committed:
flatten out subtests

we want to specify a strict ordering of operations

geoknee committed:
complete checkWithdrawal step

geoknee committed:
reorder test steps

geoknee committed:
check both ETH and token balances on l1 during withdraw

geoknee committed:
simplify callViaSafe helper

geoknee committed:
factor out expectations

geoknee commented:
In favour of #10440 .
